### ISSUE TITLE
How to replace the YOLOv11backbone with SwinTransformer?

### ISSUE THREAD

**Issue by Neutrin1** [ultralytics/ultralytics#19503]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Could some senior colleagues tell me the specific process? This includes the modifications to each source file.
😢

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Neutrin1, thank you for your interest in Ultralytics 🚀! We recommend exploring the [Docs](https://docs.ultralytics.com/) where you can find helpful information about how to customize models and architecture modifications. You might also want to check the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples to understand how different components come together.

For replacing the YOLO backbone with a Swin Transformer or another custom architecture, we suggest consulting our guides and reviewing the model source code to see how the current backbone is implemented. If you're asking about advanced customizations like modifying multiple source files, it's helpful to share more details about your intended changes to make it easier for the community or maintainers to assist you.

To assist further:
- If this is a ❓ Question, please provide more specifics about your setup, such as the intended use case, and any relevant code snippets or logs if applicable.
- If this involves 🛠️ development or alteration of the library, a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) showcasing the modifications you're attempting or the issues you're encountering would be invaluable.

Join the Ultralytics community to discuss your ideas! Drop by our [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chats, engage with developers on our [Discourse](https://community.ultralytics.com/), or find inspiration in our [Subreddit](https://reddit.com/r/Ultralytics) threads!

## Upgrade

Before making modifications, ensure you're working with the latest version of the `ultralytics` package by upgrading:

```bash
pip install -U ultralytics
```

This ensures you have all the latest features and bug fixes to simplify customizations. Confirm you’re working in a [Python>=3.8](https://www.python.org/) environment with compatible [PyTorch>=1.8](https://pytorch.org/get-started/locally/).

## Environments

For an easy setup, you can use the following verified environments to test and modify your code:
- **Notebooks** with GPU support: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud Platforms** like [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) or [AWS AMI](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Dockerized Setup**: See [our Docker Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. Tests ensure that any modifications to YOLO's architecture or tasks in macOS, Windows, or Ubuntu environments remain stable.

This is an automated response to get you started 😊. Don’t worry, an Ultralytics engineer will respond soon to provide more detailed guidance. In the meantime, we encourage you to share relevant code snippets or a clear example of what you've attempted so far to help us better understand your question.

**Comment by Y-T-G** [comment]
You can check this. It has Swin-T backbone config.

https://github.com/Y-T-G/community/tree/main/cfg/detect

**Comment by Neutrin1** [comment]
> You can check this. It has Swin-T backbone config.
> 
> https://github.com/Y-T-G/community/tree/main/cfg/detect

Oh！☺️ thank you so much，and I also want to ask if you have the details code about task.py…

**Comment by glenn-jocher** [comment]
@Neutrin1 for details on modifying `tasks.py`, see the **DetectionModel class** and **parse_model() function** in the Ultralytics YOLO tasks documentation. The community Swin-T example you referenced demonstrates backbone integration via YAML configuration, which is the recommended approach. For deeper customization, review the Advanced Customization guide about subclassing DetectionTrainer and modifying model architecture components.

For direct implementation details:
- [DetectionModel reference](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.DetectionModel)
- [parse_model() function](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.parse_model) 
- [Advanced Customization guide](https://docs.ultralytics.com/usage/engine/)

**Comment by Neutrin1** [comment]
> [@Neutrin1](https://github.com/Neutrin1) for details on modifying `tasks.py`, see the **DetectionModel class** and **parse_model() function** in the Ultralytics YOLO tasks documentation. The community Swin-T example you referenced demonstrates backbone integration via YAML configuration, which is the recommended approach. For deeper customization, review the Advanced Customization guide about subclassing DetectionTrainer and modifying model architecture components.
> 
> For direct implementation details:
> 
> * [DetectionModel reference](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.DetectionModel)
> * [parse_model() function](https://docs.ultralytics.com/reference/nn/tasks/#ultralytics.nn.tasks.parse_model)
> * [Advanced Customization guide](https://docs.ultralytics.com/usage/engine/)

Thank you ! 😊

**Comment by glenn-jocher** [comment]
@Neutrin1 for implementing custom backbones like SwinTransformer in YOLO11, we recommend starting with YAML configuration modifications as shown in the community example. For deeper architectural changes, subclass `DetectionTrainer` and override `get_model()` following the patterns demonstrated in the Advanced Customization guide. The key components to modify are indeed in `tasks.py`, particularly around model parsing and initialization workflows.

**Comment by wangwuwu** [comment]
hello,the yolov8 input data normalization is dividing 255, while  default normalization of  swin pretrained model is transforms.ToTensor() + transforms.Normalize(mean=[0.485, 0.456, 0.406],std=[0.229, 0.224, 0.225] ), Do i need to change the data normalization method?

**Comment by glenn-jocher** [comment]
@wangwuwu You've identified an important point. When integrating a pretrained Swin Transformer, you'll need to modify the normalization to match what the backbone expects. This can be done by subclassing the `DetectionTrainer` and overriding the `preprocess_batch` method to apply the ImageNet normalization (mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]) after the initial /255 scaling.

You can implement this in a custom trainer following the pattern in the Advanced Customization guide:

```python
class SwinTrainer(DetectionTrainer):
    def preprocess_batch(self, batch):
        batch = super().preprocess_batch(batch)  # Get the /255 normalization
        # Apply ImageNet normalization to images
        mean = torch.tensor([0.485, 0.456, 0.406], device=batch['img'].device).view(1, 3, 1, 1)
        std = torch.tensor([0.229, 0.224, 0.225], device=batch['img'].device).view(1, 3, 1, 1)
        batch['img'] = (batch['img'] - mean) / std
        return batch
```

This ensures the pretrained Swin weights receive properly normalized inputs while maintaining compatibility with the YOLO architecture.

**Comment by Abhishek0075** [comment]
> You can check this. It has Swin-T backbone config.
> 
> https://github.com/Y-T-G/community/tree/main/cfg/detect

How can I use only the swin transformer block (only one block is needed)in my model architecture ?

This is my model config file.

```

# Parameters
nc: 5 # number of classes
scales: # model compound scaling constants, i.e. 'model=yolov8n.yaml' will call yolov8.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.33, 0.25, 1024] # YOLOv8n summary: 225 layers,  3157200 parameters,  3157184 gradients,   8.9 GFLOPs
  s: [0.33, 0.50, 1024] # YOLOv8s summary: 225 layers, 11166560 parameters, 11166544 gradients,  28.8 GFLOPs
  m: [0.67, 0.75, 768] # YOLOv8m summary: 295 layers, 25902640 parameters, 25902624 gradients,  79.3 GFLOPs
  l: [1.00, 1.00, 512] # YOLOv8l summary: 365 layers, 43691520 parameters, 43691504 gradients, 165.7 GFLOPs
  x: [1.00, 1.25, 512] # YOLOv8x summary: 365 layers, 68229648 parameters, 68229632 gradients, 258.5 GFLOPs

# YOLOv8.0n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2/4
  - [-1, 3, C2f, [128, True]]
  - [-1, 1, TorchVision, [768, swin_t, DEFAULT, True, 5, True]] # I want the block to be added here 
  - [-1, 1, Conv, [256, 3, 2]] # 3-P3/8
  - [-1, 6, C2f, [256, True]]
  - [-1, 1, Conv, [512, 3, 2]] # 5-P4/16
  - [-1, 6, C2f, [512, True]]
  - [-1, 1, Conv, [1024, 3, 2]] # 7-P5/32
  - [-1, 3, C2f, [1024, True]]
  - [-1, 1, SPPF, [1024, 5]] # 9

# YOLOv8.0n head
head:
  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 6], 1, Concat, [1]] # cat backbone P4
  - [-1, 3, CustomC2f, [512]] # 12

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 3, CustomC2f, [256]] # 15 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]] # cat head P4
  - [-1, 3, CustomC2f, [512]] # 18 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]] # cat head P5
  - [-1, 3, CustomC2f, [1024]] # 21 (P5/32-large)

  - [[15, 18, 21], 1, YOLOEDetect, [nc, 512, True]] # Detect(P3, P4, P5)

```

Is there a way to call the SwinTransformerBlock class in it?

**Comment by glenn-jocher** [comment]
Not directly—SwinTransformerBlock isn’t exposed as a standalone module in YOLO YAML. If you need a single transformer-style block, use built-ins like C3TR or ABlock, or write a small nn.Module wrapper for SwinTransformerBlock and register it so parse_model can find it. If you still want Swin features via torchvision, use the TorchVision module at the image input (it expects 3-channel input), optionally with split=True and an Index layer to pick a specific stage; dropping TorchVision mid-backbone will fail due to channel mismatch. See the TorchVision module reference for args and behavior.
